core/schema/module.go

```diff
@@ -791,36 +791,31 @@ func (s *moduleSchema) moduleToSchemaFor(ctx context.Context, module *core.Modul
 
 		newObjResolver := ObjectResolver{}
 		for _, field := range objTypeDef.Fields {
+			field := field
+
 			fieldASTType, err := s.typeDefToSchema(field.TypeDef, false)
 			if err != nil {
 				return nil, err
 			}
-			fieldName := gqlFieldName(field.Name)
 			astDef.Fields = append(astDef.Fields, &ast.FieldDefinition{
-				Name:        fieldName,
+				Name:        field.Name,
 				Description: formatGqlDescription(field.Description),
 				Type:        fieldASTType,
 			})
 
-			// if this is an IDable type, add a resolver that converts the ID into
-			// the real object, otherwise its schema will be called against the
-			// string
-			if field.TypeDef.Kind == core.TypeDefKindObject {
-				newObjResolver[fieldName] = func(p graphql.ResolveParams) (any, error) {
-					res, err := graphql.DefaultResolveFn(p)
-					if err != nil {
-						return nil, err
-					}
-					id, ok := res.(string)
-					if !ok {
-						return nil, fmt.Errorf("expected string %sID, got %T", field.TypeDef.AsObject.Name, res)
-					}
-					return core.ResourceFromID(id)
+			fromID := s.createIDResolver(field.TypeDef, dest)
+			newObjResolver[field.Name] = func(p graphql.ResolveParams) (any, error) {
+				if src, ok := p.Source.(map[string]any); ok {
+					p.Source = map[string]any{field.Name: src[field.OriginalName]}
 				}
-			} else {
-				// no resolver to add; fields rely on the graphql "trivial resolver"
-				// where the value is just read from the parent object
-				_ = 1
+				res, err := graphql.DefaultResolveFn(p)
```

Comment:
nit: I think it might be a little safer to just change `p.Info.FieldName` to be `field.OriginalName` before passing to `graphql.DefaultResolveFn`, rather than asserting on the map type and changing it there.

Then we don't need an assumption about `p.Source` being `map[string]any` specifically. As far as I can think, that should always be the case today, but technically the graphql server lib allows it to be more than that and maybe we'd want to take advantage of that someday? It just feels like a nice assumption to remove if trivial.

If this is less easy than it sounds no worries, not a blocker.

Reply:
Yeahhh, this is entirely right and very trivial. I could have *sworn* I tried this and it didn't work - but it clearly does, so let's just use that.